### DISCUSSION TITLE
SPA -> SSR (partial hydration, whatever) in Nuxt: Potential benefits?

### DISCUSSION THREAD

**Discussion by alecStewart1** [nuxt/nuxt#29725] [Questions]
Hello, friends!

So currently we have 2 Nuxt projects, one using the default SSR/partial hydration that Nuxt already does, and one that's an SPA.

We have our reasons for one being an SPA and one not, but there's some things I've been wondering about that I thought I'd ask here. Let me give an overview.

### Overview

The SPA is a little slow in comparison, but part of that isn't Nuxt's fault. That app fetches _a lot_ of data from an external API we maintain, and the most obvious bottleneck is there an not with the Nuxt SPA, and the problems with solving the performance of the app is outside of what Vue and Nuxt can do. 

However there's a _little_ voice in the back of my head that tells me that we're not really taking advantage of Nuxt as much as we could be. Even if it doesn't speed up things on the app by much, I feel like there has be somethings we can do better to offer a better experience for our users if we got rid of the `ssr: false` in our Nuxt config.

#### Project outline

This SPA acts as our logged in site. I apologize for the vagueness of things, but obviously I don't want to reveal too much about our process.

The flow of the app is as follows

1. User logs in
2. A request to our external API is made to get a token.
3. Token is stored in a Pinia store.
4. User can then go around the site, given they have proper access (based on the permissions in the token).
5. User clicks around site.
6. Components may make requests (asynchronously) to populate relevant data within them, doing so with functions that'll call `$fetch`.
7. The data is generally stored in a Pinia store.
8. Previously mentioned functions may not actually send a request to our external API and just grab data from the respective Pinia store, should the corresponding data already be there in the store.

### Initial thoughts

One of the things I think we _might_ benefit from getting rid of `ssr: false` is using `<NuxtIsland>` where we can. Most of the SPA is interactive, there's not really much we can avoid with that, however it feels like it would make sense to use in the few places we can.

Part of me also feels like requests shouldn't be made client side in components and instead server side by Nitro, and then client side things can just be pull out of the Pinia store. I have no idea if this has any more benefits than potential downsides, but one downside being having to reorganize a lot of the parts in current application. I can't imagine having the server make requests to store data in a Pinia store would be much faster, especially since the client might have to wait on the server to finish, than just doing it client-side, but there's still that little gut feeling that we should try something different.

We are also working on providing a "loading" feature to components, as to not halt pages loading because some components might be still getting data from a request.

There are some things that have to be in a `<ClientOnly>` because of how some third-party libraries work, so part of me thinks we should try using Cloudflare's [Zaraz](https://www.cloudflare.com/application-services/products/zaraz/) as well.

### Thoughts?

I may do web development as part of my job, but I'm not a web wizard or Javascript/Typescript aficionado. In fact there's a lot of things I **don't** like about web development, but either way, I want to learn about these more advanced things to be a better developer. 

In all honesty, I'm not sure if there's much we can do performance wise on the front-end, but I thought I'd ask out of curiosity.

**Comment by xjccc** [comment]
https://nuxt.com/docs/getting-started/introduction#server-side-rendering

There has some details about server side rendering.
